File: core/src/models/sync.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{crypto::NoteCryptoMetadata, services::attachment::AttachmentCryptoMetadata};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckSyncRequest {
    #[serde(rename = "to_check")]
    pub notes: Vec<CheckNoteSyncStatus>,
    pub full_sync: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckNoteSyncStatus {
    pub local_id: String,
    pub cloud_id: Option<String>,
    pub hard_deleted: bool,
    pub cloud_version: Option<i64>,
    pub attachments: Option<Vec<AttachmentSyncCheck>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AttachmentSyncCheck {
    pub attachment_id: Uuid,
    pub checksum_encrypted: String,
    pub size_bytes: i64,
    pub is_encrypted: bool,
    pub hard_deleted: bool,
    pub file_name: String,
    pub mime_type: String,
    pub created_at: i64,
    pub updated_at: i64,
    pub crypto_metadata: Option<AttachmentCryptoMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CheckSyncResponse {
    pub to_upload: Vec<String>,
    pub notes_to_download: Vec<DownloadNote>,
    pub notes_synced: Vec<String>,
    pub attachments_to_upload: Vec<UploadAttachment>,
    pub attachments_to_download: Vec<DownloadAttachment>,
    pub attachments_synced: Vec<String>,

    #[serde(default)]
    pub attachments_to_hard_delete: Vec<String>,

    pub notes_to_hard_delete: Vec<String>,

    #[serde(default)]
    pub notes_failed: Vec<String>,

    #[serde(default)]
    pub quota_exceeded: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UploadAttachment {
    pub attachment_id: String,
    pub upload_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadNote {
    pub cloud_id: String,

    // Server-side only. This should normally never be sent by the server.
    #[serde(skip_serializing, skip_deserializing)]
    pub owner_id: String,

    pub cloud_version: i64,

    pub title: String,
    pub summary: String,
    pub content: String,

    pub created_at: i64,
    pub updated_at: i64,

    pub is_deleted: bool,
    pub deleted_at: Option<i64>,

    pub hard_deleted: bool,
    pub is_encrypted: bool,

    // Server-side only.
    #[serde(skip_serializing, skip_deserializing)]
    pub deleted_attachments: Option<Vec<String>>,

    pub crypto_meta: Option<NoteCryptoMetadata>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DownloadAttachment {
    pub attachment_id: Uuid,

    pub file_name: String,
    pub mime_type: String,
    pub size_bytes: i64,

    pub cloud_key: String,
    pub checksum_encrypted: String,

    pub is_encrypted: bool,
    pub note_cloud_id: String,
    pub crypto_meta: Option<AttachmentCryptoMetadata>,

    pub created_at: i64,
    pub updated_at: i64,

    pub download_url: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeleteAttachment {
    pub attachment_id: Uuid,
    pub cloud_key: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct NoteForUpload {
    pub local_id: String,
    pub mongo_id: Option<String>,
    pub owner_id: String,
    pub title: String,
    pub summary: String,
    pub content_path: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
    pub deleted_at: Option<i64>,
    pub version: i64,
    pub cloud_version: Option<i64>,
    pub sync_state: String,
    pub is_deleted: bool,
    pub encrypted: bool,
    pub hard_deleted: bool,
    pub crypto_meta: Option<String>,
    pub content: String,
}

#[derive(Debug)]
pub struct AttachmentForUpload {
    pub attachment_id: String,
    pub note_local_id: String,
    pub filename: String,
    pub mime_type: String,
    pub size_bytes: i64,
    pub local_path: Option<String>,
    pub cloud_key: Option<String>,
    pub checksum_encrypted: String,
    pub encrypted: bool,
    pub crypto_meta: Option<String>,
    pub sync_state: String,
    pub created_at: i64,
    pub updated_at: i64,
    pub note_cloud_id: String,
    pub content: Vec<u8>,
}

File: server/models/sync.go
package models

import (
	"time"

	"github.com/google/uuid"
	"go.mongodb.org/mongo-driver/v2/bson"
)

// **Client sends metadata of all notes that participate in sync.**
// **LocalOnly and WaitingForTombstone notes are not included.**

type CheckSyncRequest struct {
	Notes    []CheckNoteSyncStatus `json:"to_check" validate:"dive"`
	FullSync bool                  `json:"full_sync"`
}

type CheckNoteSyncStatus struct {
	LocalID     string         `bson:"local_id" json:"local_id" validate:"required"`
	SyncState   string         `json:"sync_state" validate:"required"`
	CloudID     *bson.ObjectID `bson:"_id,omitempty" json:"cloud_id,omitempty"`
	HardDeleted bool           `bson:"hard_deleted" json:"hard_deleted"`

	// **nil = note has never been uploaded**
	CloudVersion *int64 `bson:"cloud_version,omitempty" json:"cloud_version,omitempty"`

	Attachments []AttachmentSyncCheck `json:"attachments,omitempty" validate:"dive"`
}

type AttachmentSyncCheck struct {
	AttachmentID      uuid.UUID            `json:"attachment_id" validate:"required"`
	ChecksumEncrypted string               `json:"checksum_encrypted" validate:"required"`
	SizeBytes         int64                `json:"size_bytes"`
	IsEncrypted       bool                 `json:"is_encrypted"`
	FileName          string               `json:"file_name" validate:"required"`
	HardDeleted       bool                 `json:"hard_deleted"`
	MimeType          string               `json:"mime_type" validate:"required"`
	CreatedAt         int64                `bson:"created_at" json:"created_at"`
	UpdatedAt         int64                `bson:"updated_at" json:"updated_at"`
	CryptoMetadata    AttachmentCryptoMeta `json:"crypto_metadata,omitempty"`
}

type CheckSyncResponse struct {
	// **Notes that the client needs download**
	ToUpload []string `json:"to_upload"`

	// **Notes that the client needs to upload.**
	NotesToDownload []DownloadNote `json:"notes_to_download"`

	NotesSynced []string `json:"notes_synced"`

	// **Attachments that the client needs to download.**
	AttachmentsToUpload []UploadAttachment `json:"attachments_to_upload"`

	AttachmentsToDownload []DownloadAttachment `json:"attachments_to_download"`

	AttachmentsSynced []string `json:"attachments_synced"`

	AttachmentsToHardDelete []string `json:"attachments_to_hard_delete"` // **NEW**

	// **Notes that should be permanently deleted from the client.**
	NotesToHardDelete []string `json:"notes_to_hard_delete"`

	NotesFailed   []string `json:"notes_failed"`
	QuotaExceeded bool     `json:"quota_exceeded"`
}

type UploadAttachment struct {
	AttachmentId string `json:"attachment_id"`
	UploadUrl    string `json:"upload_url"`
}

type DownloadNote struct {
	CloudID      bson.ObjectID `bson:"_id" json:"cloud_id" validate:"required"`
	OwnerID      string        `bson:"owner_id" json:"-"`
	CloudVersion int64         `bson:"cloud_version" json:"cloud_version" validate:"required"`

	Title   string `bson:"title" json:"title"`
	Summary string `bson:"summary" json:"summary"`
	Content string `bson:"content" json:"content"`

	CreatedAt64 int64 `bson:"created_at" json:"created_at"`
	UpdatedAt64 int64 `bson:"updated_at" json:"updated_at"`

	IsDeleted bool   `bson:"is_deleted" json:"is_deleted"`
	DeletedAt *int64 `bson:"deleted_at,omitempty" json:"deleted_at,omitempty"`

	HardDeleted bool `bson:"hard_deleted" json:"hard_deleted"` //**on rust side when sync State = Waiting for tombstone**

	IsEncrypted bool `bson:"is_encrypted" json:"is_encrypted"`

	DeletedAttachments []string        `bson:"deleted_attachments,omitempty" json:"-"`
	CryptoMeta         *NoteCryptoMeta `bson:"crypto_meta,omitempty" json:"crypto_meta,omitempty"`
}

type DownloadAttachment struct {
	AttachmentID uuid.UUID `json:"attachment_id" validate:"required"`

	FileName  string `json:"file_name"`
	MimeType  string `json:"mime_type"`
	SizeBytes int64  `json:"size_bytes"`

	CloudKey          string `json:"cloud_key" validate:"required"` //**metadata returns to string**
	CloudNoteId       string `json:"note_cloud_id" validate:"required"`
	ChecksumEncrypted string `json:"checksum_encrypted" validate:"required"`

	IsEncrypted bool                  `json:"is_encrypted"`
	CryptoMeta  *AttachmentCryptoMeta `json:"crypto_meta,omitempty"`
	CreatedAt   int64                 `json:"created_at"`
	UpdatedAt   int64                 `json:"updated_at"`
	DownloadUrl string                `json:"download_url"`
}

type DeleteAttachment struct {
	AttachmentID uuid.UUID `json:"attachment_id" validate:"required"`
	CloudKey     string    `json:"cloud_key" validate:"required"`
}

type NoteCryptoMeta struct {
	TitleNonce   string `bson:"title_nonce" json:"title_nonce"`
	SummaryNonce string `bson:"summary_nonce" json:"summary_nonce"`
	ContentNonce string `bson:"content_nonce" json:"content_nonce"`
}

type AttachmentCryptoMeta struct {
	AttachmentNonce string `bson:"attachment_nonce" json:"attachment_nonce"`
}

type NoteUploadRequest struct {
	LocalID string  `json:"local_id" validate:"required"`
	MongoID *string `json:"mongo_id,omitempty"`

	Title   string `json:"title"`
	Summary string `json:"summary"`

	CreatedAt int64  `json:"created_at"`
	UpdatedAt int64  `json:"updated_at"`
	DeletedAt *int64 `json:"deleted_at,omitempty"`

	Version      int64  `json:"version"`
	CloudVersion *int64 `json:"cloud_version,omitempty"`

	SyncState   string `json:"sync_state"`
	IsDeleted   bool   `json:"is_deleted"`
	IsEncrypted bool   `json:"encrypted"`
	HardDeleted bool   `json:"hard_deleted"`
	// Raw JSON string, not a nested object — see note below.
	CryptoMeta *string `json:"crypto_meta,omitempty"`

	Content string `json:"content"`
}

type QuotaReservation struct {
	ID           bson.ObjectID `bson:"_id,omitempty"`
	UserID       string        `bson:"user_id"`
	AttachmentID string        `bson:"attachment_id"`
	SizeBytes    int64         `bson:"size_bytes"`
	Status       string        `bson:"status"`
	ExpiresAt    time.Time     `bson:"expires_at"`
	CreatedAt    time.Time     `bson:"created_at"`
}
